packages/tpetra/core/src/Tpetra_withLocalAccess.hpp

```diff
@@ -209,101 +253,327 @@ namespace Tpetra {
   //////////////////////////////////////////////////////////////////////
 
   /// \brief Declare that you want to access the given global object's
-  ///   local data in read-only mode, in the object's default memory
-  ///   space.
+  ///   local data in read-only mode.
   template<class GlobalObjectType>
   Details::LocalAccess<
     GlobalObjectType,
-    typename Details::DefaultMemorySpace<GlobalObjectType>::type,
-    Details::AccessMode::ReadOnly>
+    Details::read_only>
   readOnly (GlobalObjectType&);
 
   /// \brief Declare that you want to access the given global object's
   ///   local data in read-only mode (overload for const
-  ///   GlobalObjectType), in the object's default memory space.
+  ///   GlobalObjectType).
   template<class GlobalObjectType>
   Details::LocalAccess<
     GlobalObjectType,
-    typename Details::DefaultMemorySpace<GlobalObjectType>::type,
-    Details::AccessMode::ReadOnly>
+    Details::read_only>
   readOnly (const GlobalObjectType&);
 
   /// \brief Declare that you want to access the given global object's
-  ///   local data in write-only mode, in the object's default memory
-  ///   space.
+  ///   local data in write-only mode.
   template<class GlobalObjectType>
   Details::LocalAccess<
     GlobalObjectType,
-    typename Details::DefaultMemorySpace<GlobalObjectType>::type,
-    Details::AccessMode::WriteOnly>
+    Details::write_only>
   writeOnly (GlobalObjectType&);
 
   /// \brief Declare that you want to access the given global object's
-  ///   local data in read-and-write mode, in the object's default
-  ///   memory space.
+  ///   local data in read-and-write mode.
   template<class GlobalObjectType>
   Details::LocalAccess<
     GlobalObjectType,
-    typename Details::DefaultMemorySpace<GlobalObjectType>::type,
-    Details::AccessMode::ReadWrite>
+    Details::read_write>
   readWrite (GlobalObjectType&);
 
   ////////////////////////////////////////////////////////////
   // LocalAccess struct
   ////////////////////////////////////////////////////////////
 
   namespace Details {
-    /// \brief Declaration of access intent for a global object.
+    //! Tag indicating an unspecified type in LocalAccessTraits.
+    struct unspecified_type {};
+
+    /// \brief Deduce types from parameter pack of LocalAccess.
     ///
-    /// Users aren't supposed to make instances of this class.  They
-    /// should use readOnly, writeOnly, or readWrite instead, then
-    /// call instance methods like on() and valid() on the resulting
-    /// LocalAccess instance.
+    /// Deduce the following types from Args:
+    /// <ul>
+    /// <li>execution_space (Kokkos execution space)</li>
+    /// <li>memory_space (Kokkos memory space)</li>
+    /// <li>access_mode (Access type)</li>
+    /// </ul>
+    template<class ... Args>
```

Comment:
Parameter pack recursion looks like this:

1. Declare generic thing.
2. Define base case by partially specializing on zero parameters (`template<>`).
3. Define recursive case by partially specializing on `First, Rest...`.